lib/controller.ex

```diff
@@ -1,70 +1,135 @@
 defmodule Controller do
   def main(_args) do
-    run(:welcome_screen)
+    %{
+      input: :welcome,
+      header: nil,
+      content: nil,
+      view: :welcome,
+      io: IO,
+      prompt: nil,
+      menu: nil
+    }
+    |> run()
   end
 
-  def run(prompt) when prompt != "Q" do
-    UserInterface.clear_screen()
-    parse_input(prompt)
-    UserInterface.line_break()
+  def run(%{input: "Q"} = context) do
+    context
+    |> UserInterface.clear_screen()
+    UserInterface.display("Goodbye!\n")
+  end
 
-    Messages.get_prompt(:menu)
+  def run(context) do
+    context
+    |> UserInterface.clear_screen()
+    |> update_context()
+    |> Formatter.print_screen()
     |> UserInterface.get_input()
     |> run()
   end
 
-  def run("Q") do
-    execute_command("Q")
+  def update_context(context) do
+    context
+    |> update_content()
+    |> update_header()
+    |> update_menu()
+    |> update_prompt()
   end
 
-  def parse_input(input) do
-    Messages.get_prompt(input)
-    |> UserInterface.display()
+  def update_prompt(%{view: view} = context) do
+    %{context | prompt: Messages.get_prompt(view)}
+  end
 
-    execute_command(input)
+  def update_header(%{view: view} = context) do
+    %{context | header: Messages.get_header(view)}
   end
 
-  def execute_command(input) do
-    case input do
-      :welcome_screen ->
-        nil
+  def update_menu(%{view: view} = context) do
+    %{context | menu: Messages.get_menu(view)}
+  end
 
-      _ ->
-        case Integer.parse(input) do
-          :error ->
-            case input do
-              "G" ->
-                Messages.get_recipe(:ice_cubes)
+  def update_content(context) do
+    context
+    |> fetch_content()
+  end
+
+  def fetch_content(%{input: input, view: view} = context) do
```

Comment:
i think you could split this function to be smaller. maybe with the use of multiple function heads, rather than a giant case statement?

Reply:
def fetch_content(%{input: input, :welcome} = context) do
   ...
def fetch_content(%{input: input, :index } = context) do
   ...
def fetch_content(%{input: input, :view_recipe } = context) do
 ...
def fetch_content(%{input: input, :grocery_list } = context) do
 ...

That way it is matching on the arguments, rather than using nested case statements

Reply:
I love!!! Yes, will swap. Much more Elixir-y.